### PR TITLE
Fixes:Gives access of consultation details to users for only allowed facilities.

### LINKED ISSUE #8029: Modify non-admin users access to the patient consultation record
**Describe the bug**

Once a patient is discharged from a facility, the non-admin users of the facility, can view the details page and see the card of the consultation specific to their facility, but when they try to view the consultation detail page, they are taken to a 404 page

**To Reproduce**

**Expected behavior**

Refer the mentioned conversation for final behavior
(https://github.com/ohcnetwork/care_fe/pull/8727#issuecomment-2472832743)

**Screenshots**

image
image

### PR DESCRIPTION
**Pull request by Sulochan-khadka**
## Proposed Changes

- Fixes #8029 
- Change auth based access to consultation page

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

## Summary by CodeRabbit

- **New Features**
	- Enhanced button functionality in the ConsultationCard to conditionally disable actions based on facility permissions.
	- Integrated facility permission checks for improved user interaction during consultations.

- **Bug Fixes**
	- Retained error handling for the "Add Consultation Updates" button to ensure proper notifications for bed assignments.

### CODE REVIEW

**Comment by nihal467** [comment]
image

- [ ] buttons are disabled for the consultation card to which the users have access to, which is not the expected behavior, they should be allowed to access it, check the comment and issues once again @Sulochan-khadka

**Comment by Sulochan-khadka** [comment]
> image
> 
> * [ ]  buttons are disabled for the consultation card to which the users have access to, which is not the expected behavior, they should be allowed to access it, check the comment and issues once again @Sulochan-khadka

Apologies for the incorrect changes.
The issue is not very clear , which is why i asked there but was no reply, therefore i went ahead and implemented whatever i understood. It would be great of you if you kindly provide the steps to reproduce the issue.

**Comment by nihal467** [comment]
- [ ] Any non-admin users, should be able to access only the consultation card from facility there are linked, if the patient has any consultation history from facility where they are not linked, then the button on that particular card should be disabled

step 1: Admit Patient A to Facility A, then discharge them from Facility A. Repeat this process for Facility B and Facility C.
Step 2: After these admissions and discharges, Patient A will have a consultation history from Facilities A, B, and C.
Step 3: Log in with a nurse's user ID that is linked only to Facilities A and B.
Step 4: Now re-admit Patient A to one of the facilities (either A or B) where the nurse is linked.
Step 5: Access Patient A’s consultation history. In the consultation cards, the button in the consultation card related to Facility C should be disabled, as the nurse is not linked to Facility C.

**Comment by Sulochan-khadka** [comment]
> * [ ]  Any non-admin users, should be able to access only the consultation card from facility there are linked, if the patient has any consultation history from facility where they are not linked, then the button on that particular card should be disabled
> 
> step 1: Admit Patient A to Facility A, then discharge them from Facility A. Repeat this process for Facility B and Facility C. Step 2: After these admissions and discharges, Patient A will have a consultation history from Facilities A, B, and C. Step 3: Log in with a nurse's user ID that is linked only to Facilities A and B. Step 4: Now re-admit Patient A to one of the facilities (either A or B) where the nurse is linked. Step 5: Access Patient A’s consultation history. In the consultation cards, the button in the consultation card related to Facility C should be disabled, as the nurse is not linked to Facility C.

Thankyou very much for the much needed explaination. I have made some changes, kindly review them. And let me know in case of modifications.

**Comment by nihal467** [comment]
@Sulochan-khadka 
image

- [ ] Now, the district admin itself is blocked from accessing the button in consultation card, before requesting another review, make sure admins (district and state admin) have no restriction at all and non-admins (doctor, nurse, all other users) have restrictions are per issue

**Comment by Sulochan-khadka** [comment]
> @Sulochan-khadka image
> 
> * [ ]  Now, the district admin itself is blocked from accessing the button in consultation card, before requesting another review, make sure admins (district and state admin) have no restriction at all and non-admins (doctor, nurse, all other users) have restrictions are per issue

The logic seems pretty convincing that the consultation facilities are checked whether it matches with the allowed facilities of the user.

Screenshot 2024-10-15 at 1 54 11 AM

However, the problem you stated occurs when the facility is no more linked with the district admin and i think that is correct since it may happen that some facilities get linked to different district admins and therefore, the previous admin should not have the access to it. Right?!
Screenshot 2024-10-15 at 2 08 02 AM


Also "View / Upload Consultation Files" should also have same added condition right? That if the user is not allowed a facility, it should not be able to interact with the consultation records. Right?! @nihal467

**Comment by bodhish** [comment]
@Sulochan-khadka The district admins and state admins currently have a super power that they can view everything that they can see. We will eventually be taking it down. 

In this specific case lets keep it blocked, if someone comes up with an issue lets understand what they want. From a privacy perspective its always safe to limit maximum access.

@nihal467 do mark someone to address these questions when you see it.

**Comment by bodhish** [comment]
also @Sulochan-khadka add a proper title for the PR

**Comment by Sulochan-khadka** [comment]
@bodhish so my logic is fine right? Also  if the user is not allowed a facility, it should not be able to interact with the consultation records. Right?! 

Also i have changed the PR title. I hope this solves the objectives

**Comment by Sulochan-khadka** [comment]
@nihal467 @bodhish @rithviknishad is the proposed solution appropriate? Let me know, so that i can solve the merge conflicts.

**Comment by bodhish** [comment]
@nihal467

**Comment by Sulochan-khadka** [comment]
do i need to add these implementations recommended by this bot , or  is it optional? @nihal467

**Comment by nihal467** [comment]
image
image

@bodhish Currently, admins can access all routes of the disabled button in the patient details page from the patient consultation page, but this approach isn’t ideal for effectively restricting access. Many users with district admin roles rely on this access to manage certain data.

The main issue we’re aiming to resolve is to ensure that non-admin users cannot access consultation files of a facility that isn’t linked to their user ID. If a non-admin user who is not linked to a facility tries to access facility data, either by clicking on a patient card to view consultations from the patient list page or by using the "View Consultation" button on the patient details page, an error message—such as "You don't have access to this file"—should be displayed instead of redirecting them to a 404 page.

Could you review his deploy preview along with my comment, and then provide a response?

**Comment by bodhish** [comment]
Does our backend send data for these requests / pages?

**Comment by nihal467** [comment]
@bodhish 

> Does our backend send data for these requests / pages?

backend will not send the data if the user doesn't have access to the consultation record

**Comment by bodhish** [comment]
In that case should we over engineer? The page would render the empty message right 🤔

**Comment by nihal467** [comment]
> In that case should we over engineer? The page would render the empty message right 🤔

Yes, rather than disabling the button, we can simply throw an error on the frontend stating, "You don't have access to this file," replacing the current behavior of redirecting restricted users to a 404 page.

CC: @Sulochan-khadka

**Comment by nihal467** [comment]
> > In that case should we over engineer? The page would render the empty message right 🤔
> 
> Yes, rather than disabling the button, we can simply throw an error on the frontend stating, "You don't have access to this file," replacing the current behavior of redirecting restricted users to a 404 page.
> 
> CC: @Sulochan-khadka

@bodhish is this approach ok ?

File: src/Components/Facility/ConsultationCard.tsx [hunk 1]
```diff
@@ -173,6 +176,9 @@ export const ConsultationCard = (props: ConsultationProps) => {
                 `/facility/${itemData.facility}/patient/${itemData.patient}/consultation/${itemData.id}`,
               )
             }
+            disabled={
+              !facilities?.some((item) => item.name === itemData.facility_name)
```
Comment (bodhish) [inline src/Components/Facility/ConsultationCard.tsx:180]: don't we have facility id in the api ? 🤔

**Review by bodhish — CHANGES_REQUESTED** [review]

**Comment by bodhish** [comment]
Closing the issue as don't have any progress.

**Pull request closed without merging**